### DISCUSSION TITLE
[1장] 퀴즈

### DISCUSSION THREAD

**Discussion by easyhyun00** [FE-tech-talk/TechTalk-React#3] [Q&A] [answered]
## 🙋‍♀️ 문제

1. 타입스크립트에서 제너릭, any, unknown 의 간단한 특징과 차이점은?
2. 리액트에서 얕은 복사까지 구현한 이유는?
3. 리액트에서 클로저를 사용했을 때 장점

**Comment by CitrusSoda** [comment, marked as answer]
1. 제너릭 : 하나의 타입이 아닌 여러 타입에 대응할 수 있도록 만들어준다.
any : 어떠한 타입이든 상관없이 허용해준다.
unknown: 알 수 없는 타입, 사용할 때 분기 처리를 하여 사용할 수 있다.

2. Object.is를 사용하여 얕은 복사까지 가능 및 컴포넌트 프롭 내부까지 비교할 필요 없음(?)

3. 변수 은닉화 및 독립 시행 가능

**Comment by easyhyun00** [comment]
## 1번 답변

**1. 타입스크립트에서 제너릭, any, unknown 의 간단한 특징과 차이점은?**

`제너릭` - 다양한 타입을 유연하게 처리하며, 타입 안정성 유지
(그래서 고차함수 보면 제너릭 타입으로 되어있음)

`any` - 모든 타입을 허용, 타입 안전성 포기
(타입스크립트의 정적 타이핑 이점을 모두 버리기 때문)

`unknown` - 모든 타입을 허용, 타입 검사를 필요로 하며 any 대신 unknown 사용 권장

**Comment by easyhyun00** [comment]
## 2번 답변

**2. 리액트에서 얕은 복사까지 구현한 이유는?**

첫 번째 깊이까지만 수행하는데 내부 객체까지 완벽하게 비교하기 위해 재귀문을 사용하면 객체 안에 객체가 몇 개까지 있을지 알 수 없기 때문에 리액트 성능에 악영향을 미침

즉, 성능상 이유와 얕은 복사를 사용하면 변경된 부분만 쉽게 추적 가능

**Comment by easyhyun00** [comment]
## 3번 답변

**3. 리액트에서 클로저를 사용했을 때 장점**

리액트에서 함수 컴포넌트의 구조, 작동 방식, 훅의 원리, 의존성 배열 등 클로저에 의존

클로저를 활용하면 전역 스코프 사용 막고, 원하는 정보만 노출시킬 수 있음

useState를 사용하여 React 함수 컴포넌트 훅에서 클로저를 사용

함수 내부 상태 유지, 외부에서 접근 못해서 캡슐화 모듈화 강화, 불필요한 렌더링 방지(클로저 통해 필요한 함수만 재사용)
